File: shared/utils/materiality_bands.py
"""Canonical materiality bands.

Materiality is a 1.0-10.0 significance scale (see the white paper and
MATERIALITY_SCORE_PROMPT). These bands are the single shared bucketing used
across the API, report pipeline, and agent tools — symbolic activity is a
real soft-power signal (narrative projection, signaling), not noise, so the
platform surfaces all three bands rather than only ranking by score.

Client-side mirror: client/src/components/materialityBands.ts
"""
from __future__ import annotations

# Band boundaries: [lower, upper) — upper bound exclusive except the last.
SYMBOLIC_MAX = 4.0      # 1.0-3.9: visits, statements, ceremonies, gestures
DEVELOPING_MAX = 7.0    # 4.0-6.9: MOUs, pilots, training — commitments not delivery
                        # 7.0-10:  funded projects, signed deals, delivered hardware

BANDS = ("symbolic", "developing", "substantive")


def band_for(score: float | None) -> str | None:
    """Return the band name for a materiality score, or None if unscored."""
    if score is None:
        return None
    if score < SYMBOLIC_MAX:
        return "symbolic"
    if score < DEVELOPING_MAX:
        return "developing"
    return "substantive"


def band_bounds(band: str) -> tuple[float | None, float | None]:
    """(min_inclusive, max_exclusive) score bounds for a band name."""
    if band == "symbolic":
        return (None, SYMBOLIC_MAX)
    if band == "developing":
        return (SYMBOLIC_MAX, DEVELOPING_MAX)
    if band == "substantive":
        return (DEVELOPING_MAX, None)
    raise ValueError(f"unknown materiality band: {band!r}")


# SQL CASE fragment classifying a score column into a band label.
def band_sql(column: str = "material_score") -> str:
    return (
        f"CASE WHEN {column} IS NULL THEN NULL "
        f"WHEN {column} < {SYMBOLIC_MAX} THEN 'symbolic' "
        f"WHEN {column} < {DEVELOPING_MAX} THEN 'developing' "
        f"ELSE 'substantive' END"
    )

File: client/src/components/materialityBands.ts
/**
 * Canonical materiality bands — client mirror of shared/utils/materiality_bands.py.
 *
 * Materiality is a 1.0-10.0 significance scale. Symbolic activity is a real
 * soft-power signal (narrative projection, signaling), so the UI gives each
 * band its own identity instead of treating low scores as "less important".
 */

export type MaterialityBand = 'symbolic' | 'developing' | 'substantive'

export const SYMBOLIC_MAX = 4.0
export const DEVELOPING_MAX = 7.0

export function bandFor(score: number | null | undefined): MaterialityBand | null {
  if (score == null) return null
  if (score < SYMBOLIC_MAX) return 'symbolic'
  if (score < DEVELOPING_MAX) return 'developing'
  return 'substantive'
}

export const BAND_LABELS: Record<MaterialityBand, string> = {
  symbolic: 'Symbolic',
  developing: 'Developing',
  substantive: 'Substantive',
}

// Badge colors: substantive = green, developing = amber, symbolic = blue.
// Blue (not gray) on purpose — symbolic gestures are signal, not noise.
export const BAND_BADGE: Record<MaterialityBand, { bg: string; fg: string }> = {
  symbolic: { bg: '#dbeafe', fg: '#1e40af' },
  developing: { bg: '#fef3c7', fg: '#92400e' },
  substantive: { bg: '#dcfce7', fg: '#166534' },
}

// Chart series colors per band (solid, chart-friendly)
export const BAND_COLORS: Record<MaterialityBand, string> = {
  symbolic: '#3b82f6',
  developing: '#f59e0b',
  substantive: '#16a34a',
}

export function badgeStyleFor(score: number | null | undefined): { backgroundColor: string; color: string } {
  const band = bandFor(score)
  if (!band) return { backgroundColor: '#f1f5f9', color: '#475569' }
  return { backgroundColor: BAND_BADGE[band].bg, color: BAND_BADGE[band].fg }
}
